reports/tasks/msisdn_message_report.py

```diff
@@ -13,63 +16,200 @@ class GenerateMSISDNMessageReport(BaseTask):
 
-            identity = results[0]
+            data[msisdn] = {'id': results[0]['id'],
+                            'created_at': results[0]['created_at']}
 
-            sheet.add_row({
-                'Phone number': msisdn,
-                'Date registered': identity['created_at'],
+        return data
+
+    def retrieve_registration_info(self, is_client, data):
+        logger = self.get_logger()
+
+        for msisdn in data:
+            if data[msisdn].get('id', None) is None:
+                # Skip if we didn't find an identity
+                continue
+
+            # Currently we'll only be working with mother msisdns
+            registration = Registration.objects.filter(
+                    mother_id=data[msisdn]['id']
+                ).order_by('-created_at').first()
+
+            if registration is None:
+                logger.info(
+                    'No registration found with mother_id {0} ({1})'
+                    .format(data[msisdn]['id'], msisdn))
+                continue
+
+            data[msisdn]['reg_date'] = registration.created_at.strftime(
+                "%Y-%m-%d %H:%M:%S")
+            data[msisdn]['msg_type'] = registration.data.get('msg_type', "")
+            data[msisdn]['preg_week'] = registration.data.get('preg_week', "")
+
+            # Get facility info from the operator's identity
+            operator_id = registration.data.get('operator_id', None)
+            if operator_id is not None:
+                operator_identity = is_client.get_identity(operator_id)
+                data[msisdn]['facility'] = operator_identity.get(
+                        'details', {}).get('facility_name', "")
+            else:
+                logger.info(
+                    'No operator_id on registration for {0}'.format(msisdn))
+                data[msisdn]['facility'] = ""
+
+        return data
+
+    def retrieve_messages(self, ms_client, data, start_date, end_date):
+        logger = self.get_logger()
+
+        longest_list = 0
+        for msisdn in data:
+            message_list = []
+
+            if data[msisdn].get('id', None) is None:
+                # Skip if we didn't find an identity
+                continue
+
+            response = ms_client.get_outbounds({
+                "to_identity": data[msisdn]['id'],
+                "after": start_date.strftime("%Y-%m-%dT00:00:00"),
+                "before": end_date.strftime("%Y-%m-%dT00:00:00")
             })
+            results = list(response['results'])
+
+            if len(results) < 1:
+                logger.info(
+                    'No results from message sender for {0}'.format(msisdn))
+
+                data[msisdn]["messages"] = message_list
+                continue
+
+            for message in results:
+                message_list.append({
+                    "content": message['content'],
+                    # Reformat the date
+                    "date_sent": message['created_at'].replace('T', ' ')[:-8],
```

Comment:
This seems a little fragile to me. Maybe it's better to first get the string into python's `datetime` format, and then to `strftime` it to get it into the format that we want.